File: leet_code/trees/build_tree_post_inorder.rb
class TreeNode

  attr_accessor :val, :left, :right, :next
  def initialize(val)
    @val = val
    @left = nil
    @right = nil
    @next = nil
  end
end


def build_tree(inorder, postorder)
    return nil if inorder.empty?

    idx = inorder.index(postorder[-1])
    new_node = TreeNode.new(postorder[-1])

    new_node.left = build_tree(inorder[0...idx], postorder[0...idx])
    new_node.right = build_tree(inorder[idx+1..-1], postorder[idx...-1])
    new_node
end

File: leet_code/trees/build_tree_post_inorder.js
class TreeNode{
    constructor(val){
      this.val = val;
      this.left = null;
      this.right = null;
      this.next = null;
    }
}


const buildTree = (inorder, postorder) => {
    if (inorder.length === 0) return null;

    const inLength = inorder.length - 1,
          postLength = postorder.length -1,
          idx = inorder.indexOf(postorder[postLength]),
          newNode = new TreeNode(postorder[postLength]);

    newNode.left = buildTree(inorder.slice(0,idx), postorder.slice(0,idx));
    newNode.right = buildTree(inorder.slice(idx+1,inLength+1), postorder.slice(idx, postLength));
    return newNode;
};
